### ISSUE TITLE
L19

### ISSUE THREAD

**Issue by uzinny04** [uzinny04/AI_Appl#5]
Long Short Term Memory
Inports

import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim

# for reproducibility
torch.manual_seed(100)
     
<torch._C.Generator at 0x7f816f1319d0>
1 Data Construction

# Dictionary
sample_sentences = ['howareyou', 'whats up?', 'iamgreat.']
char_set = list(set(''.join(sample_sentences)))
dic = {c: i for i, c in enumerate(char_set)}

# Parameters
dic_size = len(dic)
input_size = dic_size
hidden_size = dic_size * 2
output_size = dic_size

# Dataset setting
input_batch = []
target_batch = []

for sentence in sample_sentences:
  x_data = [dic[c] for c in sentence[:-1]]
  x_one_hot = [np.eye(dic_size)[x] for x in x_data]
  y_data = [dic[c] for c in sentence[1:]]
  input_batch.append(x_one_hot)
  target_batch.append(y_data)

# To torch tensors
X = torch.FloatTensor(np.array(input_batch))
Y = torch.LongTensor(np.array(target_batch))

print(X.shape)
print(Y.shape)
     
torch.Size([3, 8, 17])
torch.Size([3, 8])
2 RNN model
Input (입력의 형태)
Input type: torch.Tensor
Input shape: (N x S x E)
N: Batch size, S: Sequence length, E: Embedding size
Hidden (출력의 형태)
Hidden type: torch.Tensor

# Model
class Custom_RNN(torch.nn.Module):
  def __init__(self, input_dim, hidden_dim, output_dim, layers):
    super(Custom_RNN, self).__init__()
    self.rnn = torch.nn.RNN(input_dim, hidden_dim, num_layers=layers)
    self.fc = torch.nn.Linear(hidden_dim, output_dim, bias=True)

  def forward(self, x):
    x, _status = self.rnn(x)
    x = self.fc(x)
    return x

learning_rate = 0.05
training_epochs = 100
model = Custom_RNN(input_size, hidden_size, output_size, 2)
     
3 Train

# define cost/loss & optimizer
criterion = nn.CrossEntropyLoss()    # Softmax
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

# train
for epoch in range(training_epochs):
  optimizer.zero_grad()
  outputs = model(X)
  loss = criterion(outputs.reshape(-1, dic_size), Y.reshape(-1))
  loss.backward()
  optimizer.step()
  if epoch % 10 == 9:
    print('epoch: ',epoch, 'loss: ', loss.item())

result = outputs.data.numpy().argmax(axis=2)
for sentence in result:
  print(''.join([char_set[c] for c in np.squeeze(sentence)]))

     
epoch:  9 loss:  0.09620252251625061
epoch:  19 loss:  0.058530014008283615
epoch:  29 loss:  0.05801613628864288
epoch:  39 loss:  0.05786002054810524
epoch:  49 loss:  0.057837288826704025
epoch:  59 loss:  0.05780743435025215
epoch:  69 loss:  0.05780068039894104
epoch:  79 loss:  0.05779645964503288
epoch:  89 loss:  0.0577949695289135
epoch:  99 loss:  0.05779321491718292
ouareyou
hats up?
amgreat.

3 Assignment
다음 미완성 코드를 활용해 좀 더 긴 문장을 학습해보자
- Sample sentences

"if you want to build a ship, don't drum up people together to "
"collect wood and don't assign them tasks and work, but rather "
"teach them to long for the endless immensity of the sea."
Training data sentence

Shape: (N, S, E)
Hidden

Shape: (N, S, E *2)
Output

Shape: (N, S, E)

import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim

# for reproducibility
torch.manual_seed(100)

# Dictionary
sample_sentence_1 = "if you want to build a ship, don't drum up people together to "
sample_sentence_2 = "collect wood and don't assign them tasks and work, but rather "
sample_sentence_3 = "teach them to long for the endless immensity of the sea."
sample_sentence = sample_sentence_1 + sample_sentence_2 + sample_sentence_3
char_set = list(set(sample_sentence))
dic = {c: i for i, c in enumerate(char_set)}

# Parameters
dic_size = len(dic)
input_size = dic_size  # One-hot encoded input size
hidden_size = dic_size * 2
output_size = dic_size
unit_sequence_length = 20
num_layers = 2 # Number of RNN layers for deeper structure

# Dataset setting
input_batch = []
target_batch = []

for i in range(0, len(sample_sentence) - unit_sequence_length):
    # 입력 시퀀스: 현재 문자부터 L 길이
    X_seq = sample_sentence[i:i + unit_sequence_length]
    # 타겟 시퀀스: 다음 문자부터 L 길이 (입력 시퀀스보다 한 칸 앞당겨짐)
    Y_seq = sample_sentence[i+1:i + unit_sequence_length + 1]

    # 문자를 인덱스로 변환
    X_idx = [dic[c] for c in X_seq]
    Y_idx = [dic[c] for c in Y_seq]

    # X_idx를 원-핫 인코딩
    X_one_hot = np.zeros((unit_sequence_length, dic_size), dtype=np.float32)
    for j, idx in enumerate(X_idx):
        X_one_hot[j, idx] = 1.0

    input_batch.append(X_one_hot)
    target_batch.append(Y_idx)

# To torch tensors
X = torch.FloatTensor(np.array(input_batch)) # Shape: (Batch, Seq Length, Input Size)
Y = torch.LongTensor(np.array(target_batch)) # Shape: (Batch, Seq Length)

print(f"Total number of batches: {X.shape[0]}")
print(f"Sequence Length: {X.shape[1]}")
print(f"Dictionary Size: {dic_size}")
print(X.shape)
print(Y.shape)

# Model
class Custom_RNN(nn.Module):

    def __init__(self, input_size, hidden_size, output_size, num_layers=1, cell_type='lstm'):
        super().__init__()
        self.cell_type = cell_type
        self.hidden_size = hidden_size
        self.num_layers = num_layers
        self.output_size = output_size

        # RNN Cell 정의
        if cell_type == 'rnn':
            self.rnn_cell = nn.RNN(input_size, hidden_size, num_layers, batch_first=True)
        elif cell_type == 'gru':
            self.rnn_cell = nn.GRU(input_size, hidden_size, num_layers, batch_first=True)
        elif cell_type == 'lstm':
            self.rnn_cell = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True)
        else:
            raise ValueError("Unsupported cell_type. Must be 'rnn', 'lstm', or 'gru'.")

        # Hidden state를 최종 출력 크기로 매핑하는 FC layer
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        # x shape: (Batch Size, Sequence Length, Input Size)

        # 초기 Hidden State 및 Cell State (LSTM의 경우)
        h0 = torch.zeros(self.num_layers, x.size(0), self.hidden_size).to(x.device)

        if self.cell_type == 'lstm':
            c0 = torch.zeros(self.num_layers, x.size(0), self.hidden_size).to(x.device)
            out, _ = self.rnn_cell(x, (h0, c0))
        else:
            out, _ = self.rnn_cell(x, h0)

        # out shape: (Batch Size, Sequence Length, Hidden Size)

        # 모든 시퀀스 단계의 출력에 FC layer 적용을 위해 모양 변경 (B*L, H)
        out = out.reshape(-1, self.hidden_size)

        # 최종 출력 (B*L, Output Size)
        out = self.fc(out)

        # 최종 출력 모양 (B, L, Output Size)로 복원 (선택적)
        out = out.reshape(x.size(0), x.size(1), self.output_size)

        return out

learning_rate = 0.05
training_epochs = 100
# cell_type='lstm'을 기본으로 사용합니다. ('rnn' 또는 'gru'로 변경 가능)
model = Custom_RNN(input_size, hidden_size, output_size, num_layers=num_layers, cell_type='lstm')

# define cost/loss & optimizer
criterion = nn.CrossEntropyLoss()    # Softmax
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

# train
for epoch in range(training_epochs):
  optimizer.zero_grad()
  # outputs shape: (Batch, Seq Length, Output Size)
  outputs = model(X)

  # CrossEntropyLoss를 위해 shape을 (B*L, Output Size)와 (B*L)로 변경
  loss = criterion(outputs.reshape(-1, dic_size), Y.reshape(-1))

  loss.backward()
  optimizer.step()

  if (epoch + 1) % 10 == 0:
    # 10 에포크마다 loss 출력 (0부터 시작하므로 9 대신 10의 배수 사용)
    print(f'Epoch: {epoch+1:03d}, Loss: {loss.item():.4f}')

# result
# outputs.data.numpy().argmax(axis=2) shape: (Batch, Seq Length)
results = outputs.data.numpy().argmax(axis=2)

predicted_sentences = []
for batch_result in results:
  # 각 배치별로 인덱스를 문자로 변환
  predicted_sentence = ''.join([char_set[c] for c in np.squeeze(batch_result)])
  predicted_sentences.append(predicted_sentence)

# 전체 문장 길이와 비교를 위한 예측된 문자열 (각 배치의 첫 번째 결과만 이어 붙이는 방식)
# 실제 전체 문장을 재구성하는 것은 더 복잡한 추론 로직이 필요하나, 여기서는 학습된 배치의 결과만 출력합니다.
print('\n--- Predicted Sequence Samples ---')
for i in range(5):
    if i < len(predicted_sentences):
        print(f'Batch {i+1} Prediction: "{predicted_sentences[i]}"')
    else:
        break

# 정확도 계산 (전체 시퀀스에 대한 문자 일치율)
total_predictions = Y.numel()
correct_predictions = (torch.LongTensor(results) == Y).sum().item()
accuracy = correct_predictions / total_predictions * 100

print(f'\nOriginal Sentence Length: {len(sample_sentence)}')
print(f'Total Predictions Checked: {total_predictions}')
print(f'Accuracy (Character Match): {accuracy:.2f}%')

print('\nOriginal Target Sequence (First Batch):')
print(''.join([char_set[c.item()] for c in Y[0]]))
torch.Size([160, 20, 25])
torch.Size([160, 20])
epoch:  9 loss:  2.7191596031188965
epoch:  19 loss:  1.671645164489746
epoch:  29 loss:  0.6607065796852112
epoch:  39 loss:  0.27085548639297485
epoch:  49 loss:  0.17204250395298004
epoch:  59 loss:  0.14130252599716187
epoch:  69 loss:  0.12837830185890198
epoch:  79 loss:  0.12267328053712845
epoch:  89 loss:  0.11952614784240723
epoch:  99 loss:  0.11799737066030502
p you want to build a ship, don't drum up people together to collect wood and don't assign them tasks and work, but rather teach them to long for the endless immensity of the sea.
Accuracy =  0.994413407821229
